src/main/docs/guide/appendix/architecture/iocArch.adoc

```diff
@@ -0,0 +1,65 @@
+Micronaut is an implementation of the JSR-330 specification for Dependency Injection.
+
+Dependency Injection (or Inversion of Control) is a widely adopted and common pattern in Java that allows loosely decoupling components to allow applications to be easily extended and tested. 
+
+The way in which objects are wired together is decoupled from the objects themselves in this model by a separate programming model. In the case of Micronaut this model is based annotations defined within the JSR-330 specification plus an extended set of annotations located within the https://docs.micronaut.io/latest/api/io/micronaut/context/annotation/package-summary.html[io.micronaut.context.annotation] package.
+
+These annotations are visited by the <<compilerArch, Micronaut Compiler>> which traverses the source code language AST and builds a model used to wire objects together at runtime. 
+
+NOTE: It is important to note that the actual object wiring is deferred until runtime.
+
+For Java code api:annotation.processing.BeanDefinitionInjectProcessor[] (which is a Java Annotation https://docs.oracle.com/en/java/javase/17/docs/api/java.compiler/javax/annotation/processing/Processor.html[Processor]) is invoked from the Java compiler for each class annotated with a bean definition annotation.
+
+NOTE: What constitutes a bean defining annotation is complex as it takes into account meta-annotations, but in general it is any annotation annotated with a JSR-330 bean `@Scope`
+
+The `BeanDefinitionInjectProcessor` will visit each bean in the user code source and generate additional byte code using the ASM byte code generation library that sits along side the annotated class in the same package.
+
+NOTE: For historic reasons the dependency injection processor does not use the api:inject.visitor.TypeElementVisitor[] API but will likely do so in the future
+
+Byte code generation is implemented in the api:inject.writer.BeanDefinitionWriter[] which contains methods to "visit" different aspects of the way is bean is defined (the api:inject.BeanDefinition[]).
+
+The following diagram illustrates the flow:
+
+image::arch/beanwriter.png[]
+
+For example given the following type:
+
+snippet::io.micronaut.docs.inject.intro.Vehicle[tags="class",indent=0]
+
+The following is generated:
+
+* A `example.$Vehicle$Definition$Reference` class that implements the api:inject.BeanDefinitionReference[] interface that allows the application to soft load the bean definition without loading all of the metadata or the class itself (in the case where the introspected class is itself not on the classpath). Since references are loaded via link:{jdkapi}/java/util/ServiceLoader.html[ServiceLoader] An entry in a generated `META-INF/services/io.micronaut.inject.BeanDefinitionReference` referring to this type is also generated at compilation time..
```

Comment:
```suggestion
* A `example.$Vehicle$Definition$Reference` class that implements the api:inject.BeanDefinitionReference[] interface that allows the application to soft load the bean definition without loading all of the metadata or the class itself (in the case where the introspected class is itself not on the classpath). Since references are loaded via link:{jdkapi}/java/util/ServiceLoader.html[ServiceLoader] an entry in a generated `META-INF/services/io.micronaut.inject.BeanDefinitionReference` referring to this type is also generated at compilation time.
```